### PR TITLE
feat: add TxStatus rpc endpoint

### PR DESCRIPTION
**Pull request by ninabarbakadze**
## Description

Fixes - #1163

we've had some back and forth re this feature, and have decided to implement this simple rpc endpoint as an interim solution while we figure out and design an efficient way of querying transaction status during its life cycle.

### CODE REVIEW

File: proto/tendermint/store/types.proto [hunk 1]
```diff
@@ -7,3 +7,9 @@ message BlockStoreState {
   int64 base   = 1;
   int64 height = 2;
 }
+
+message TxIndex {
+  int64 height    = 1;
+  int64 index     = 2;
+  bool  committed = 3;
```
Comment (cmwaters) [inline proto/tendermint/store/types.proto:14]: I would save the tx code that gets returned in TxResponse (I think it's uint32)

File: store/store.go [hunk 2]
```diff
@@ -283,8 +301,9 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
 
 	pruned := uint64(0)
 	batch := bs.db.NewBatch()
+	batchTxs := bs.db.NewBatch()
```
Comment (cmwaters) [inline store/store.go:304]: Why do you create a different batch. You should use the same batch in order for the transaction to be atomic

File: store/store.go [hunk 3]
```diff
@@ -322,20 +345,41 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
...
+			txValue, err := bs.db.Get(txKey)
+			if err != nil {
+				return 0, err
+			}
+			if txValue != nil {
+				if err := batchTxs.Delete(txKey); err != nil {
+					return 0, err
+				}
+			}
+		}
```
Comment (cmwaters) [inline store/store.go:362]: This will need to be done prior to deleting the block else there won't be any block to query

File: store/store.go [hunk 4]
```diff
@@ -445,6 +500,17 @@ func (bs *BlockStore) SaveSeenCommit(height int64, seenCommit *types.Commit) err
...
+func (bs *BlockStore) SaveTxIndex(txHash []byte, txIndex *cmtstore.TxIndex) error {
+	bz, err := proto.Marshal(txIndex)
+	if err != nil {
+		panic(fmt.Errorf("error serializing txIndex: %v", err))
+	}
+	if err := bs.db.Set(calcTxHashKey(txHash), bz); err != nil {
+		panic(err)
+	}
+	return nil
+}
```
Comment (cmwaters) [inline store/store.go:512]: Ideally we batch this as well

Comment (cmwaters) [reply store/store.go:512]: But don't worry about the performance too much if we're planning on switching the storage engine in the near future

File: store/store.go (hunk 2 above)
Comment (ninabarbakadze) [reply store/store.go:304]: first batch is for the blocks second one is for txs

File: proto/tendermint/store/types.proto [hunk 5]
```diff
@@ -7,3 +7,8 @@ message BlockStoreState {
   int64 base   = 1;
   int64 height = 2;
 }
+
+message TxStatus {
+  int64 height    = 1;
+  int64 index     = 2;
+}
```
Comment (rootulp) [inline proto/tendermint/store/types.proto:14]: [nit] this seems more like `TxLocation`  because it identifies where a transaction exists in a block but it doesn't contain any info about the status of the transaction. Should we rename the type?

File: store/store.go [hunk 6]
```diff
@@ -513,6 +556,23 @@ func LoadBlockStoreState(db dbm.DB) cmtstore.BlockStoreState {
...
+	if err != nil {
+		panic(err)
+	}
+	if len(bz) == 0 {
+		return nil
+	}
+
+	var txs cmtstore.TxStatus
+	if err = proto.Unmarshal(bz, &txs); err != nil {
+		panic(fmt.Errorf("unmarshal to TxIndex failed: %w", err))
```
Comment (rootulp) [inline store/store.go:571]: fwiw I like the `TxIndex` or `TxLocation` name better than `TxStatus` but I'm not strongly opinionated if we're eventually going to add more fields to `TxStatus` then that name prob makes more sense. Regardless, I think this instance failed to update after a name change:

```suggestion
		panic(fmt.Errorf("unmarshal to TxStatus failed: %w", err))
```

File: store/store_test.go [hunk 7]
```diff
@@ -375,6 +379,40 @@ func TestBlockStoreSaveLoadBlock(t *testing.T) {
 	}
 }
 
+func TestSaveBlockIndexesTxs(t *testing.T) {
+	// Create a state and a block store
+	state, blockStore, cleanup := makeStateAndBlockStore(log.NewTMLogger(new(bytes.Buffer)))
+	defer cleanup()
+
+	// Create a block with 1000 txs
```
Comment (rootulp) [inline store/store_test.go:387]: [question] does the following code make a block with 1000 txs or 1000 blocks? Since each loop execution calls `makeBlock`, I think the latter. If so:

```suggestion
	// Create 1000 blocks
```

It looks like `makeTxs` always creates 10 transactions so pretty sure the comment is incorrect

**Review by rootulp — APPROVED** [review]
Overall code LGTM! I like the tests. One question:
1. Are we planning on adding more fields to `TxStatus` at a later point? It seems more like a pointer to a location than a status.

**Comment by cmwaters** [comment]
> Are we planning on adding more fields to TxStatus at a later point? It seems more like a pointer to a location than a status.

Yes, most likely we will have a Status enum, where the transaction is in one of several states:

- Pending
- Rejected (This could be both included and rejected and excluded and rejected)
- Evicted
- Committed

We might also want to include an error code.

The goal with this endpoint is to provide sufficient real-time observability of the transaction.

**Review by cmwaters — COMMENTED** [review]
Few minor comments else otherwise looks good

File: store/store_test.go (hunk 7 above)
Comment (ninabarbakadze) [reply store/store_test.go:387]: it's supposed to be 1000 blocks, txs was by mistake

**Review by evan-forbes — APPROVED** [review]
dope, upstream might want this

File: proto/tendermint/store/types.proto (hunk 5 above)
Comment (staheri14) [reply proto/tendermint/store/types.proto:14]: [suggestion] Could we also have a comment about what `index` represents?

File: rpc/core/blocks.go [hunk 8]
```diff
@@ -179,6 +179,17 @@ func BlockByHash(ctx *rpctypes.Context, hash []byte) (*ctypes.ResultBlock, error
 	return &ctypes.ResultBlock{BlockID: blockMeta.BlockID, Block: block}, nil
 }
 
+// TxStatus retrieves the status of a transaction given its hash. It returns a ResultTxStatus
+// containing the height and index of the transaction within the block.
+func TxStatus(hash []byte) (*ctypes.ResultTxStatus, error) {
```
Comment (staheri14) [inline rpc/core/blocks.go:184]: An explanation of the components of the hash would be beneficial. For example, is this related to the transaction key? What fields are included in the digest?

File: proto/tendermint/store/types.proto (hunk 5 above)
Comment (staheri14) [reply proto/tendermint/store/types.proto:14]: I also agree with Rootul's point that `TxStatus` contains information pertains to a transaction already included in one of the previous blocks. However, the term `Status` seems to imply the current state of a transaction that is still being processed.

File: store/store.go [hunk 9]
```diff
@@ -471,6 +506,10 @@ func calcBlockHashKey(hash []byte) []byte {
 	return []byte(fmt.Sprintf("BH:%x", hash))
 }
 
+func calcTxHashKey(hash []byte) []byte {
```
Comment (staheri14) [inline store/store.go:509]: Renaming this function would be helpful since it only converts the hash to a string, without performing any calculations. It could be given a more accurate name, or alternatively, the operation could be carried out inline instead of using a separate function for it.

File: store/store.go [hunk 10]
```diff
@@ -513,6 +552,23 @@ func LoadBlockStoreState(db dbm.DB) cmtstore.BlockStoreState {
 	return bsj
 }
 
+// LoadTxStatus retrieves the status of a transaction from the block store given its hash.
+func (bs *BlockStore) LoadTxStatus(txHash []byte) *cmtstore.TxStatus {
+	bz, err := bs.db.Get(calcTxHashKey(txHash))
+	if err != nil {
+		panic(err)
```
Comment (staheri14) [inline store/store.go:559]: Is panic essential at this point? I think it is perceivable that someone request tx status for a wrong/unavailable hash.

**Review by staheri14 — COMMENTED** [review]
It looks good, however, there are a few aspects that could use some clarification and improvement before final approval.

File: store/store.go (hunk 9 above)
Comment (ninabarbakadze) [reply store/store.go:509]: I'm following a general pattern of generating keys in blockstore. view https://github.com/celestiaorg/celestia-core/blob/1d70a8e659d6dc1dee5057ba47bd0d92f99d0e59/store/store.go#L454

File: store/store.go (hunk 10 above)
Comment (ninabarbakadze) [reply store/store.go:558]: i think for wrong/unavailable hash we'd get nil

File: proto/tendermint/store/types.proto [hunk 11]
```diff
@@ -7,3 +7,8 @@ message BlockStoreState {
   int64 base   = 1;
   int64 height = 2;
 }
+
+message TxInfo {
+  int64 height    = 1;
+  int64 index     = 2;
+}
```
Comment (rootulp) [inline proto/tendermint/store/types.proto:14]: [not blocking][optional] docs would be nice.

```suggestion
// TxInfo describes the location of a tx inside a committed block.
message TxInfo {
  int64 height    = 1;
  int64 index     = 2;
}
```

File: rpc/core/mempool.go [hunk 12]
```diff
@@ -59,6 +59,7 @@ func BroadcastTxSync(ctx *rpctypes.Context, tx types.Tx) (*ctypes.ResultBroadcas
 	}
 }
 
+// DEPRECATED: Use BroadcastTxSync or BroadcastTxAsync alongside TxStatus instead.
```
Comment (rootulp) [inline rpc/core/mempool.go:62]: [question] what do you mean by alongside `TxStatus`? I would expect users to invoke `BroadcastTxSync` or `BroadcastTxAsync` and then subsequently invoke `TxStatus`. Maybe we can remove `alongside TxStatus` here and add to the Go doc of `BroadcastTxSync` and `BroadcastTxAsync` that users can invoke `TxStatus` after.

```suggestion
// DEPRECATED: Use BroadcastTxSync or BroadcastTxAsync instead.
```

File: rpc/core/types/responses.go [hunk 13]
```diff
@@ -60,6 +60,11 @@ type ResultCommit struct {
 	CanonicalCommit    bool `json:"canonical"`
 }
 
+type ResultTxStatus struct {
```
Comment (rootulp) [inline rpc/core/types/responses.go:63]: [nit] docs

```suggestion
// ResultTxStatus contains info to locate a tx in a committed block.
type ResultTxStatus struct {
```

**Review by rootulp — APPROVED** [review]

File: store/store.go (hunk 9 above)
Comment (staheri14) [reply store/store.go:508]: Oh, I see. Then, that's an unusual naming convention in my opinion. Though, nothing to be done about it in this PR.

File: store/store.go (hunk 10 above)
Comment (staheri14) [reply store/store.go:558]: > i think for wrong/unavailable hash we'd get nil

Correct, I am proposing the removal of the panic behavior. Alternatively, could you explain why retaining the panic option is beneficial?

Comment (ninabarbakadze) [reply store/store.go:558]: usually, an error returned from the db is a severe issue, like disk corruption. blockstore generally panics if a read/write operation on the database results in an error, and I'm trying to follow that pattern.

If we want to change this behaviour and return an error instead, we can discuss a potential refactor of the other getters as well. However, i'd say that's beyond the scope of this pr.

**Review by evan-forbes — APPROVED** [review]

**Review by rootulp — APPROVED** [review]

File: store/store.go (hunk 10 above)
Comment (staheri14) [reply store/store.go:558]: I understand now. Since the existence of the hash in the previous blocks is already verified in another part of the code, this function will never be invoked with a non-existing hash. All good then!

**Review by staheri14 — APPROVED** [review]
LGTM!
